Add release instructions (#57)
Azure/PyRIT: doc/contributing.md

## Releasing PyRIT to PyPI

Create an account on pypi.org if you don't have one yet.
Ask one of the other maintainers to add you to the `pyrit` project on PyPI.

Make sure the version data in pyproject.toml is set correctly.
We use standard versioning for Python.
Below, we use x.y.z as the example version.

Before running the following command replace all "local" links like
"./doc/README.md" with links that will work from any website, i.e.,
"https://github.com/Azure/PyRIT/blob/releases/vx.y.z/doc/README.md"

Commit your changes and push them to the repository on a branch called

> `releases/vx.y.z`, then run

Reviewer: How are we doing the versioning? This might be a separate convo but just curious :) We should provide guidance on this if this is for all contributors

Author: It's for maintainers mostly. Non-maintainers don't have permissions to do releases. https://py-pkgs.org/07-releasing-versioning.html

Reviewer: Ah! If this is just for maintainers, lets add that. It makes sense but I wasn't sure on who had which permissions.

> ## Releasing PyRIT to PyPI

Reviewer: When do we submit a new PyRIT release? Is this with every PR?

Author: No, that would be too much work. There are two approaches: periodically (e.g., once a month), or with significant changes (e.g., a crucial bugfix, or a big new feature). We haven't really settled on either one yet.

> 

Reviewer: ```suggestion
Only maintainers have permissions to create a new release.
```

Author: Oh I see this just now. Added something similar.

Create an account on pypi.org if you don't have one yet.

> Ask one of the other maintainers to add you to the `pyrit` project on PyPI.

Reviewer: Will folks know who the maintainers are? Is there a link we can add here as a reference of where to find maintainers?

Author: They're listed in pyproject.toml

> Make sure the version data in pyproject.toml is set correctly.

Reviewer: How do we know if version data is set correctly?

Author: It should not be the same as the previous release version 😆 If you want to do a release you typically have a new version in mind in accordance with https://py-pkgs.org/07-releasing-versioning.html

Keep that version in sync with `__init__.py` which is usually set to

> the next version (e.g., x.y.z+1) with suffix `.dev0`.

Author: x.y.z+1 is confusing. Need to clarify here

Reviewer: You could add reference to https://py-pkgs.org/07-releasing-versioning.html to clarify, maybe :)

For the release branch, we can remove this suffix.
We use standard versioning for Python.
Below, we use x.y.z as the example version.

> Before running the following command replace all "local" links like

Reviewer: Why do we do this? Also will links be reset to "local" links after a release goes through or is this just for the changeset within the release?

Author: Because the README gets picked up by PyPI and PyPI doesn't pick up your other files, so local links break.

This only applies on the release branch, so not to main.

Reviewer: I'd appreciate adding this reasoning after L172 :)

Reviewer: I tried adding a suggestion for this -- resolving this thread

> Before running the following command replace all "local" links like

Author: Looks like @rdheekonda changed this in #58 but we still need to adjust all the links to point to the release branch rather than the main branch.
E.g., https://github.com/Azure/PyRIT/blob/releases/vx.y.z/ instead of https://github.com/Azure/PyRIT/blob/main/

"./doc/README.md" with links that will work from any website, i.e.,
"https://github.com/Azure/PyRIT/blob/releases/vx.y.z/doc/README.md"

> 

Reviewer: ```suggestion
This is required for the release branch because PyPI does not pick up other files besides the README, which results in local links breaking.
```

Author: Added locally!

## Releasing PyRIT to PyPI

This section is for maintainers only.
If you don't know who the maintainers are but you need to reach them
please file an issue or (if it needs to remain private) contact the
email address listed in pyproject.toml

First, decide what the next release version is going to be.
We follow semantic versioning for Python projects; see
https://semver.org/ for more details.
Below, we refer to the version as `x.y.z`.
`x` is the major version, `y` the minor version, and `z` the patch version.
Every Python project starts at `0.1.0`.
Patches such as bugfixes without breaking changes increase the patch version.
Importantly, they are backward compatible, so upgrading from `0.1.0` to
`0.1.1` (or higher ones like `0.1.38`) should not break your code.
More significant changes such as major features.

> They should still be backwards compatible, so if you're upgrading from

Reviewer: ```suggestion
More significant changes, such as major features, should still be backwards compatible, so....
```

Reviewer: I'm still not sure how to properly comment a suggestion hahaha and have been using markdown, apologies if it doesn't get the right lines for these next few nits

Author: You can add more lines! Just add a new line in your suggestion.

Author: I must have lost my train of thought there... let me rewrite, because I didn't intend for it to be what your suggestion says.

`1.1.0` to `1.2.0` your code shouldn't break.

> In reality, the major version `1.0.0` is the first "stable" release.

Reviewer: nit: remove "In reality," and start with "The major version"

Anything before (i.e., leading with major version `0`) indicates that it is
not stable and anything may change at any time.

> For that reason, we the minor version may indicate breaking changes, too,

Reviewer: "For that reason, the minor version..." --> need to remove extra "we"

## Releasing PyRIT to PyPI

This section is for maintainers only.
If you don't know who the maintainers are but you need to reach them
please file an issue or (if it needs to remain private) contact the
email address listed in pyproject.toml

> First, decide what the next release version is going to be.

Reviewer 2: do we want to also introduce a release notes file that summarizes the changes or do folks prefer to utilize git history for that? I personally would prefer the former as git history can be a bit messy.

Author: Oh that's a great question that I forgot to raise!!! In the past, I've had release notes files in the repo, but they tend to be tricky to maintain. [Sidenote: there's the question of what changes are worth mentioning, too... or should one mention everything?]

GitHub has a neat little "release notes" feature. I can show it in tomorrow's sync.

We follow semantic versioning for Python projects; see
https://semver.org/ for more details.
Below, we refer to the version as `x.y.z`.
`x` is the major version, `y` the minor version, and `z` the patch version.
Every Python project starts at `0.1.0`.

> Patches such as bugfixes without breaking changes increase the patch version.

Reviewer 2: are patches always bug fixes?

Author: semver.org says

> MAJOR version when you make incompatible API changes
> MINOR version when you add functionality in a backward compatible manner
> PATCH version when you make backward compatible bug fixes

So I read this as "yes"

Reviewer 2: right now it reads as patches are bug fixes and something else. I recommend to just say 'backward compatible bug fixes'
